File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        stack<int> stack;
        ListNode *slow=head,*fast=head;
        while(fast && fast->next){
            stack.push(slow->val);
            slow=slow->next;
            fast=fast->next?fast->next->next:fast->next;
        }   
        if(fast) slow=slow->next; // odd list 
        while(slow && stack.top()==slow->val){
            slow=slow->next;
            stack.pop();
        }
        return stack.empty();
    }
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # find the mid point using slow,fast 
        # from the mid start reverse ll then you can traverse to check if not palindrom 
        slow,fast = head,head 
        while fast: 
            slow = slow.next 
            fast = fast.next.next if fast.next else fast.next 

        prev,curr,nextnode = None,slow,None   
        while curr : 
            nextnode = curr.next 
            curr.next = prev 
            prev = curr 
            curr=nextnode 
        
        l,r = head,prev 
        while l and r: 
            if l.val!=r.val:
                return False 
            l=l.next
            r=r.next 

        return True 

        